File: ios/Classes/TencentMapApi.swift
import QMapKit
import Flutter

class _TencentMapApi: NSObject, TencentMapApi {
  let tencentMap: TencentMap
  let mapView: QMapView

  init(_ tencentMap: TencentMap) {
    self.tencentMap = tencentMap
    self.mapView = tencentMap.mapView
  }

  func setMapType(type: MapType) throws {
    mapView.mapType = [
      MapType.dark: QMapType.dark,
      MapType.normal: QMapType.standard,
      MapType.satellite: QMapType.satellite,
    ][type] ?? QMapType.standard
  }

  func setCompassEnabled(enabled: Bool) throws {
    mapView.showsCompass = enabled
  }

  func setScaleControlsEnabled(enabled: Bool) throws {
    mapView.showsScale = enabled
  }

  func setRotateGesturesEnabled(enabled: Bool) throws {
    mapView.isRotateEnabled = enabled
  }

  func setScrollGesturesEnabled(enabled: Bool) throws {
    mapView.isScrollEnabled = enabled
  }

  func setZoomGesturesEnabled(enabled: Bool) throws {
    mapView.isZoomEnabled = enabled
  }

  func setTiltGesturesEnabled(enabled: Bool) throws {
    mapView.isOverlookingEnabled = enabled
  }

  func setIndoorViewEnabled(enabled: Bool) throws {
    mapView.setIndoorEnabled(enabled)
  }

  func setTrafficEnabled(enabled: Bool) throws {
    mapView.showsTraffic = enabled
  }

  func setBuildingsEnabled(enabled: Bool) throws {
    mapView.showsBuildings = enabled
  }

  func setMyLocationButtonEnabled(enabled: Bool) throws {
    throw FlutterError(code: "505", message: "Not supported on iOS platform.", details: nil)
  }

  func setMyLocationEnabled(enabled: Bool) throws {
    mapView.showsUserLocation = enabled
  }

  func setMyLocation(location: Location) throws { }

  func setMyLocationStyle(style: MyLocationStyle) throws { }

  func moveCamera(position: CameraPosition, duration: Int64) throws {
    let animated = duration > 0
    if let it = position.target?.coordinate { mapView.setCenter(it, animated: animated) }
    if let it = position.zoom { mapView.setZoomLevel(CGFloat(it), animated: animated) }
    if let it = position.tilt { mapView.setOverlooking(CGFloat(it), animated: animated) }
    if let it = position.bearing { mapView.setRotation(CGFloat(it), animated: animated) }
  }

  func addMarker(options: MarkerOptions) throws -> String {
    var id = UUID()
    while(tencentMap.hasKeyInMarkers(key: id)) {
      id = UUID()
    }
    let annotation = options.annotation
    tencentMap.markers[id] = annotation
    mapView.addAnnotation(annotation)
    return id.uuidString
  }

  func addPolyline(options: PolylineOptions) throws -> String {
    return ""
  }

  func pause() throws { }

  func resume() throws { }

  func stop() throws { }

  func start() throws { }

  func destroy() throws { }
}

File: android/src/main/kotlin/com/morbit/tencent_map/TencentMapApi.kt
package com.morbit.tencent_map

import com.tencent.tencentmap.mapsdk.maps.CameraUpdateFactory
import com.tencent.tencentmap.mapsdk.maps.TencentMap.*

class _TencentMapApi(private val tencentMap: TencentMap) : TencentMapApi {
  private val mapView = tencentMap.view

  override fun setMapType(type: MapType) {
    mapView.map.mapType = when (type) {
      MapType.NORMAL -> MAP_TYPE_NORMAL
      MapType.SATELLITE -> MAP_TYPE_SATELLITE
      MapType.DARK -> MAP_TYPE_DARK
    }
  }

  override fun pause() {
    mapView.onPause()
  }

  override fun resume() {
    mapView.onResume()
  }

  override fun stop() {
    mapView.onStop()
  }

  override fun start() {
    mapView.onStart()
  }

  override fun destroy() {
    mapView.onDestroy()
  }

  override fun setCompassEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isCompassEnabled = enabled
  }

  override fun setRotateGesturesEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isRotateGesturesEnabled = enabled
  }

  override fun setScrollGesturesEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isScrollGesturesEnabled = enabled
  }

  override fun setZoomGesturesEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isZoomGesturesEnabled = enabled
  }

  override fun setSkewGesturesEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isTiltGesturesEnabled = enabled
  }

  override fun setIndoorViewEnabled(enabled: Boolean) {
    mapView.map.setIndoorEnabled(enabled)
  }

  override fun setIndoorPickerEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isIndoorLevelPickerEnabled = enabled
  }

  override fun setTrafficEnabled(enabled: Boolean) {
    mapView.map.isTrafficEnabled = enabled
  }

  override fun setBuildingsEnabled(enabled: Boolean) {
    mapView.map.showBuilding(enabled)
  }

  override fun setScaleControlsEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isScaleViewEnabled = enabled
  }

  override fun setMyLocationButtonEnabled(enabled: Boolean) {
    mapView.map.uiSettings.isMyLocationButtonEnabled = enabled
  }

  override fun setMyLocationEnabled(enabled: Boolean) {
    mapView.map.isMyLocationEnabled = enabled
  }

  override fun setUserLocationType(type: UserLocationType) {
    mapView.map.setMyLocationStyle(type.toMyLocationStyle())
  }

  override fun getUserLocation(): Location {
    return mapView.map.myLocation.toLocation()
  }

  override fun moveCamera(position: CameraPosition, duration: Long) {
    val cameraPosition = position.toCameraPosition(mapView.map.cameraPosition)
    val cameraUpdate = CameraUpdateFactory.newCameraPosition(cameraPosition)
    if (duration > 0) {
      mapView.map.stopAnimation()
      mapView.map.animateCamera(cameraUpdate, duration, null)
    } else {
      mapView.map.moveCamera(cameraUpdate)
    }
  }

  override fun addMarker(options: MarkerOptions): String {
    val marker = mapView.map.addMarker(options.toMarkerOptions(tencentMap.binding))
    tencentMap.markers[marker.id] = marker
    return marker.id
  }

  override fun addPolyline(options: PolylineOptions): String {
    TODO("Not yet implemented")
  }
}
